patches/server/0223-PlayerLaunchProjectileEvent.patch

```diff
@@ -5,267 +5,255 @@ Subject: [PATCH] PlayerLaunchProjectileEvent
              itemstack.shrink(1);
          }
-+        */ // Paper end
  
-         return InteractionResultHolder.sidedSuccess(itemstack, world.isClientSide());
-     }
 diff --git a/src/main/java/net/minecraft/world/item/EnderpearlItem.java b/src/main/java/net/minecraft/world/item/EnderpearlItem.java
-index c7d4745aed77b23562cde7c68b8870fa239428d4..749ab72edc0d2e9c6f1161415ab8d59d3d6ca976 100644
+index c7d4745aed77b23562cde7c68b8870fa239428d4..d67a6936dfc5e5d041ab6458a8ba4146eade2d60 100644
 --- a/src/main/java/net/minecraft/world/item/EnderpearlItem.java
 +++ b/src/main/java/net/minecraft/world/item/EnderpearlItem.java
-@@ -25,7 +25,20 @@ public class EnderpearlItem extends Item {
+@@ -20,17 +20,29 @@ public class EnderpearlItem extends Item {
+         ItemStack itemstack = user.getItemInHand(hand);
+ 
+         // CraftBukkit start - change order
++        boolean shouldConsume = true; // Paper
+         if (!world.isClientSide) {
+             ThrownEnderpearl entityenderpearl = new ThrownEnderpearl(world, user);
  
              entityenderpearl.setItem(itemstack);
              entityenderpearl.shootFromRotation(user, user.getXRot(), user.getYRot(), 0.0F, 1.5F, 1.0F);
 -            if (!world.addFreshEntity(entityenderpearl)) {
 +            // Paper start
-+                com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) user.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Projectile) entityenderpearl.getBukkitEntity());
-+            if (event.callEvent() && world.addFreshEntity(entityenderpearl)) {
-+                if (event.shouldConsume() && !user.getAbilities().instabuild) {
-+                    itemstack.shrink(1);
-+                } else if (user instanceof net.minecraft.server.level.ServerPlayer) {
-+                    ((net.minecraft.server.level.ServerPlayer) user).getBukkitEntity().updateInventory();
-+                }
-+
-+                world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), SoundEvents.ENDER_PEARL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (net.minecraft.world.entity.Entity.SHARED_RANDOM.nextFloat() * 0.4F + 0.8F));
-+                user.awardStat(Stats.ITEM_USED.get(this));
-+                user.getCooldowns().addCooldown(this, 20);
-+            } else {
++            com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) user.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Projectile) entityenderpearl.getBukkitEntity());
++            if (!event.callEvent() || !world.addFreshEntity(entityenderpearl)) {
++                user.getCooldowns().removeCooldown(this);
```

Comment:
Why do we need to remove the cooldown here? I don't think the client adds to its own cooldown.

Reply:
The client seems to predict the cooldown and still run it at least visually.

Reply:
I wonder if we should just send the packet itself then, instead of relying on that logic. Like what if we put an event call inside removeCooldown?

Reply:
That's fair, updated to just send the packet